Flash Attention for Neuron

apoorvtintin opened the pull request:
This PR adds support for flash attention kernel for Neuron implemented through Neuron Kernel Interface (NKI). 

The flash attention kernel works with TRN1 and TRN2.

apivovarov commented on axlearn/common/flash_attention/neuron_attention.py:
```diff
@@ -0,0 +1,129 @@
+from functools import partial
+import jax
+import jax.numpy as jnp
+import jax.numpy as jnp
```
dup

kelvin-zou commented on axlearn/common/flash_attention/neuron_attention.py:
```diff
@@ -0,0 +1,124 @@
+    return out
+
+
+def _mha_forward(query, key, value, bias, causal, softmax_scale):
```
Can we get a support for segment ID and dropout as well? Both are quite needed nowadays.

apoorvtintin replied:
Segment ID support is in progress and will be added soon

kelvin-zou commented on axlearn/common/flash_attention/utils.py:
```diff
@@ -159,6 +159,21 @@ def jit_attn(query, key, value, bias, segment_ids):
         return jit_attn
 
+    elif backend == "neuron":
+        from axlearn.common.flash_attention.neuron_attention import (
```
On demand import is kind of risky, we can live with it for functions inside neuron_attention.py, can we at least get it as a header import for files outside of neuron_attention.py?

kelvin-zou commented on axlearn/common/flash_attention/neuron_attention.py:
```diff
@@ -0,0 +1,124 @@
+    k = key.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, kv_seq_len]
+    v = value.transpose(0, 2, 1, 3)  # [batch_size, num_heads, kv_seq_len, d_model]
+
+    import neuronxcc.nki.language as nl
```
Please add pylint here

ruomingp reviewed (commented):
Will defer to @kelvin-zou for approval.

apoorvtintin commented:
This has been implemented in a new PR https://github.com/apple/axlearn/pull/939. That PR addresses all comments in this PR but is from a different fork of Axlearn.

apoorvtintin commented:
This has been implemented in a new PR https://github.com/apple/axlearn/pull/939. That PR addresses all comments in this PR but is from a different fork of Axlearn.
